### ISSUE TITLE
ERR_HTTP_HEADERS_SENT caused by asynchronous hooks

### ISSUE THREAD

**Issue by johnspackman** [fastify/fastify#4908]
### Prerequisites

- [X] I have written a descriptive issue title
- [X] I have searched existing issues to ensure the bug has not already been reported

### Fastify version

4.19.2

### Node.js version

20.3.0

### Operating system

macOS

### Operating system version (i.e. 20.04, 11.3, 10)

13.4.1

### Description

We frequently get this error:

```
Unhandled promise rejection: Error [ERR_HTTP_HEADERS_SENT]: Cannot write headers after they are sent to the client
    at new NodeError (node:internal/errors:405:5)
    at ServerResponse.writeHead (node:_http_server:342:11)
    at onSendEnd (/snipped-path/node_modules/fastify/lib/reply.js:572:9)
    at wrapOnSendEnd (/snipped-path/node_modules/fastify/lib/reply.js:538:5)
    at next (/snipped-path/node_modules/fastify/lib/hooks.js:224:7)
    at handleResolve (/snipped-path/node_modules/fastify/lib/hooks.js:241:5)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
```

The problem is triggered when one of my plugins has an asynchronous `onSendHook`, this is the code:

```
      fastify.addHook("onSend", (request, reply, payload) => {
        console.log("in onSend");
        let promise = new Promise(resolve => {
          setTimeout(() => {
            console.log("onSend: payload=" + payload);
            resolve(payload);
          }, 1);
        });
        return promise;
      });
```

My code handles the request with this code:

```
      app.get(
        "*",
        async (req, reply) => await this._defaultUrlHandler(req, reply)
      );
```
and then:
```
    async _defaultUrlHandler(req, reply, url) {
        reply.send("hello");
        return;
    }
```

My `_defaultUrlHandler` is called async, and it does call `reply.send()` - but `reply.sent` is never set to `true`.

Moments later, in `fastify/lib/wrapThenable` in the `wrapThenable` function, the code says:

```
function wrapThenable (thenable, reply) {
  thenable.then(function (payload) {
    if (reply[kReplyHijacked] === true) {
      return
    }

    // this is for async functions that are using reply.send directly
    //
    // since wrap-thenable will be called when using reply.send directly
    // without actual return. the response can be sent already or
    // the request may be terminated during the reply. in this situation,
    // it require an extra checking of request.aborted to see whether
    // the request is killed by client.
    if (payload !== undefined || (reply.sent === false && reply.raw.headersSent === false && reply.request.raw.aborted === false)) {
      // we use a try-catch internally to avoid adding a catch to another
      // promise, increase promise perf by 10%
      try {
        reply.send(payload)
```

The payload is `undefined` because that was what what returned by my `_defaultUrlHandler` and `reply.sent` is `false`.  This triggers the *second* call to `reply.send` which is the cause of the error.

The call stack looks like this:

<img width="375" alt="image" src="https://github.com/fastify/fastify/assets/129568/5ab52bc6-50d9-44b3-a00d-c65056e0afeb">

If I set a breakpoint in `reply.send`, when it is called for the second time, this is in the console:
```
in onSend
```

Note that the `in onSend` message has appeared, but the async function in onSend has not resolved, otherwise we would see additional messages in the console.

The problem is that in `lib/handleRequest.js` the `preHandlerCallback` calls my URL handler and receives a promise, so uses `wrapThenable`; my async URL handler calls `reply.send` but the async `onSendHook` delays the writing of the reply.  My URL handler resolves, `wrapThenable` decides that nothing has been sent (because `reply.sent === false`), and then `wrapThenable` calls `reply.send` for a second time.

This is the complete output of the console:

```
Webserver started on http://localhost:8088
in onSend
in onSend
onSend: payload=hello
onSend: payload=undefined
Unhandled promise rejection: Error [ERR_HTTP_HEADERS_SENT]: Cannot write headers after they are sent to the client
    at new NodeError (node:internal/errors:405:5)
    at ServerResponse.writeHead (node:_http_server:342:11)
    at onSendEnd (/snipped-path/node_modules/fastify/lib/reply.js:572:9)
    at wrapOnSendEnd (/snipped-path/node_modules/fastify/lib/reply.js:538:5)
    at next (/snipped-path/node_modules/fastify/lib/hooks.js:224:7)
    at handleResolve (/snipped-path/node_modules/fastify/lib/hooks.js:241:5)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
```

Note how the onSendHook is called twice, and so is the resolved promise - once with the correct payload and once with `undefined` (thanks to the `wrapThenable`).

## Solution (well, workaround / hack)
The problem seems to be that `wrapThenable` relies on `reply.sent` to know if data has been passed to `reply.send`, but that property is defined as:

```
sent: {
    enumerable: true,
    get () {
      // We are checking whether reply was hijacked or the response has ended.
      return (this[kReplyHijacked] || this.raw.writableEnded) === true
    },
```

and so `reply.sent` actually means that "the reply has finished sending to the client" and *not* "there is data which will is being sent"

A quick fix is to call `.hijack`:

```
Reply.prototype.send = function (payload) {
  if (this[kReplyIsRunningOnErrorHook] === true) {
    throw new FST_ERR_SEND_INSIDE_ONERR()
  }

  if (this.sent) {
    this.log.warn({ err: new FST_ERR_REP_ALREADY_SENT() }, 'Reply already sent')
    return this
  }
  this.hijack(); // workaround for asynchronous `onSendHook`
```

### Steps to Reproduce

see above

### Expected Behavior

`reply.send` should not be called for a second time

**Comment by mcollina** [comment]
Thanks for reporting! This behavior is well-tested and there is likely something else at play.

Can you provide _complete_ steps to reproduce? We often need a [reproducible example](https://stackoverflow.com/help/minimal-reproducible-example), e.g. some code that allows someone else to recreate your problem by just copying and pasting it. If it involves more than a couple of different files, create a new repository on GitHub and add a link to that.

**Comment by johnspackman** [comment]
I have already tried but there are a number of complex timing issues at play which means that it is hard to reproduce.

Please can you explain why `wrapThenable` would not call reply.send for a second time?  EG if `reply.sent` is `false`, which it will be?

**Comment by mcollina** [comment]
> I have already tried but there are a number of complex timing issues at play which means that it is hard to reproduce.

You have a bug there, not in Fastify. More likely you are forking the promise chain somewhere and not `return reply` at the end of your handler.

**Comment by johnspackman** [comment]
Reproduced it with plain node:

```
const fastify = require("fastify");

const app = fastify({ logger: true });

app.addHook("onSend", (request, reply, payload) => {
  console.log("in onSend");
  let promise = new Promise(resolve => {
    setTimeout(() => {
      console.log("payload=" + payload);
      resolve(payload);
    }, 1);
  });
  return promise;
});

async function _defaultUrlHandler(req, reply) {
  reply.send("hello");
  return;
}

app.get("*", async (req, reply) => await _defaultUrlHandler(req, reply));

try {
  app.listen(8088);
} catch (err) {
  app.log.error(err);
  process.exit(1);
}
```

Console output is:

```
$ node test.js 
(node:16637) [FSTDEP011] FastifyDeprecation: Variadic listen method is deprecated. Please use ".listen(optionsObject)" instead. The variadic signature will be removed in `fastify@5`.
(Use `node --trace-warnings ...` to show where the warning was created)
{"level":30,"time":1689590494403,"pid":16637,"hostname":"JohnsM2.local","msg":"Server listening at http://[::1]:8088"}
{"level":30,"time":1689590494406,"pid":16637,"hostname":"JohnsM2.local","msg":"Server listening at http://127.0.0.1:8088"}
{"level":30,"time":1689590503479,"pid":16637,"hostname":"JohnsM2.local","reqId":"req-1","req":{"method":"GET","url":"/snipped-url.html","hostname":"snipped-hostname:8088","remoteAddress":"127.0.0.1","remotePort":64065},"msg":"incoming request"}
in onSend
in onSend
payload=hello
payload=undefined
{"level":30,"time":1689590503493,"pid":16637,"hostname":"JohnsM2.local","reqId":"req-1","res":{"statusCode":200},"responseTime":13.66712498664856,"msg":"request completed"}
node:internal/errors:496
    ErrorCaptureStackTrace(err);
    ^

Error [ERR_HTTP_HEADERS_SENT]: Cannot write headers after they are sent to the client
    at new NodeError (node:internal/errors:405:5)
    at ServerResponse.writeHead (node:_http_server:342:11)
    at onSendEnd (/snippedpath/node_modules/fastify/lib/reply.js:574:9)
    at wrapOnSendEnd (/snippedpath/node_modules/fastify/lib/reply.js:540:5)
    at next (/snippedpath/www.spar.co.uk/node_modules/fastify/lib/hooks.js:224:7)
    at handleResolve (/snippedpath/www.spar.co.uk/node_modules/fastify/lib/hooks.js:241:5)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5) {
  code: 'ERR_HTTP_HEADERS_SENT'
}

Node.js v20.3.0
```

**Comment by Eomm** [comment]
don't mix async/callback style:

https://fastify.dev/docs/latest/Reference/Routes/#async-await

```diff
async function _defaultUrlHandler(req, reply) {
+  return reply.send("hello");
-  return;
}
```

**Comment by johnspackman** [comment]
> don't mix async/callback style:
> 
> https://fastify.dev/docs/latest/Reference/Routes/#async-await
> 
> ```diff
> async function _defaultUrlHandler(req, reply) {
> +  return reply.send("hello");
> -  return;
> }
> ```

The code takes a completely different path if the handler returns an instance of `Reply`?  And there is no error checking or warnings????  That's insane.  It's taken days to track this down into a reproducable example.

Sorry but this isn't the first time that if there is a problem with fastify it is a nightmare to figure out what's gone wrong.

**Comment by mcollina** [comment]
> And there is no error checking or warnings????

What would you have expected in this case? In essence, you tried to send a response twice and got an error.

The Error that was quite explicative: you tried to send the response twice.

The misuse of promises causes this. Take a look at the following videos to learn more:

- https://www.youtube.com/watch?v=XV-u_Ow47s0&t=1s&pp=ygUbYnJva2VuIHByb21pc2VzIGphbWVzIHNuZWxs
- https://www.youtube.com/watch?v=yRyfr1Qcf34&pp=ygUbYnJva2VuIHByb21pc2VzIGphbWVzIHNuZWxs

Maybe we should we add a warning in the logs pointing to the [documentation](https://fastify.dev/docs/latest/Reference/Routes/#async-await)?

>  That's insane.

Please moderate your language. You had help within hours.

>  It's taken days to track this down into a reproducible example.

According to the times reported by GitHub, it took you roughly 1 hour to write a reproducible example - considering that the original issue _did not_ include one.

**Comment by Eomm** [comment]
> You came back with "More likely you are ... not return reply at the end of your handler" pretty quick, so it seems like this is a common problem?

I must say that mixing `async` and `reply.send()` is a common thing.
Here we tried to give a clear DO/DO NOT overview: https://fastify.dev/docs/latest/Reference/Routes/#promise-resolution

I discussed it on my Fastify book too.
Rule of thumbs: forget about `reply.send()`. It will save you the 99% of times.

This is a typical node.js race condition and it is quite impossible to make everyone happy (who want to return `undefined`, who want a warning etc..)

I think the more elegant solution is:
- to develop an eslint plugin
- rewrite the fastify quick-start doc

The latter is already on my todo-list. I tried the former but it requires too much effort to do it alone.
